File: leetcode/48.cpp
// rotate 再 reverse。
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        
        for (int i = 0; i < n; i++) { // each col
            for (int j = i + 1; j < n; j++) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        
        for (int i = 0; i < n; i++)
            reverse(matrix[i].begin(), matrix[i].end());
    }
};

class Solution {
public:
    void rotate(vector<vector<int>>& m) {
        int n = m.size();
        
        for (int i = 0; i < n / 2; i++) {
            for (int j = i; j < n - i - 1; j++) {
                
                int r = i, c = j;
                int tmp = m[r][c];
                do {
                    int nr = c;
                    int nc = n - 1 - r;
                    swap(tmp, m[nr][nc]);
                    r = nr;
                    c = nc;
                } while (r != i || c != j);
            }
        }
    }
};

File: leetcode/48.rs
use std::mem::swap;

impl Solution {
    fn map(i: usize, j: usize, n: usize) -> (usize, usize) {
        (j, n - 1 - i)
    }

    pub fn rotate(mat: &mut Vec<Vec<i32>>) {
        let n = mat.len();
        'o: for i in 0..(n / 2) {
            for j in i..(n - i - 1) {
                let (mut ni, mut nj) = (i, j);
                let mut v = mat[i][j];
                loop {
                    let (_ni, _nj) = Self::map(ni, nj, n);
                    ni = _ni;
                    nj = _nj;
                    swap(&mut mat[ni][nj], &mut v);
                    if ni == i && nj == j {
                        break;
                    }
                }
            }
        }
    }
}
